File: MomentumiOS/MomentumiOS/Presentation/Transactions/TransactionViewModel.swift
//
//  TransactionViewModel.swift
//  MomentumiOS
//
//  Created by Mwai Banda on 7/18/22.
//  Copyright © 2022 Momentum. All rights reserved.
//

import Foundation
import MomentumSDK
import TinyDi

class TransactionViewModel: ObservableObject {
    @Inject private var controller: TransactionController
    @Published var transactions = [MomentumTransaction]()
    
  
    func addTransaction(
        description: String,
        date: String,
        amount: Double,
        isSeen: Bool
    ) {
        controller.addTransaction(
            description: description,
            date: date,
            amount: amount,
            isSeen: isSeen
        )
    }
    
    private func addTransactions(transactions: [MomentumTransaction]) {
        controller.addTransactions(transactions: transactions)
    }
    
     func getTransactions(userId: String) {
//         controller.deleteAllTransactions()
        controller.getTransactions(userId: userId) { [self] res in
            if let res = res.data {
                transactions = (res as! [Transaction]).map({ $0.toMomentumTransaction() })
                if !transactions.isEmpty {
                    addTransactions(transactions: transactions)
                }
            }
        }
    }
    
     
    func postTransactionInfo(transaction: Transaction, onCompletion: @escaping () -> Void = {}) {
        controller.postTransactionInfo(transaction: transaction) { response in
            if let response = response.data {
                if response == 200 {
                    Log.d(tag: "Pay/Post/Status", message: response)
                    onCompletion()
                }
            } else if let error = response.message {
                Log.d(tag: "Pay/Failure", message: error)
            }
        }
    }
    func getMomemtumTransactions(userId: String) {
        controller.getMomentumTransactions { [unowned self] transactions in
            self.transactions = transactions
            if transactions.isEmpty {
                getTransactions(userId: userId)
            }
        }
    }
    
    func deleteTransactionById(transactionId: Int) {
        controller.deleteTransactionById(transactionId: Int32(transactionId))
    }
    
    func deleteAllTransactions() {
        controller.deleteAllTransactions()
    }
    
    func getTransactionDate() -> String {
        let date = Date()
        let dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "MMMM d"
        return dateFormatter.string(from: date)
    }
}

File: MomentumAndroid/src/main/java/com/mwaibanda/momentum/android/presentation/offer/transaction/TransactionViewModel.kt
package com.mwaibanda.momentum.android.presentation.offer.transaction

import android.util.Log
import androidx.lifecycle.ViewModel
import com.mwaibanda.momentum.data.db.MomentumTransaction
import com.mwaibanda.momentum.domain.controller.TransactionController
import com.mwaibanda.momentum.domain.models.Transaction
import com.mwaibanda.momentum.utils.DataResponse
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

class TransactionViewModel(
    private val transactionController: TransactionController
): ViewModel() {

    private val _transactions: MutableStateFlow<List<MomentumTransaction>> = MutableStateFlow(emptyList())
    val transactions: StateFlow<List<MomentumTransaction>> = _transactions.asStateFlow()


    private fun getTransactions(userId: String) {
        transactionController.getTransactions(userId = userId) { res ->
            when (res) {
                is DataResponse.Failure -> {
                    Log.d("Transaction/GetFailure", res.message ?: "")
                }
                is DataResponse.Success -> {
                    _transactions.value = res.data?.map { it.toMomentumTransaction() } ?: emptyList()
                    if (_transactions.value.isNotEmpty()) {
                        addTransactions(transactions = _transactions.value)
                    }
                    Log.d("Transaction/GetSuccess", "${_transactions.value}")
                }
            }
        }
    }
    fun getMomentumTransactions(userId: String) {
        transactionController.getMomentumTransactions { transactions ->
            _transactions.value = transactions
            if (transactions.isEmpty()) {
                getTransactions(userId = userId)
            }
        }
    }

    fun postTransactionInfo(transaction: Transaction, onCompletion: () -> Unit) {
        transactionController.postTransactionInfo(transaction = transaction) { res ->
            when (res) {
                is DataResponse.Failure -> {
                    Log.d("Pay/PostFailure", res.message ?: "")
                }
                is DataResponse.Success -> {
                    onCompletion()
                    Log.d("Pay/PostSuccess", res.message ?: "")
                }
            }
        }

    }

    fun addTransaction(
        description: String,
        date: String,
        amount: Double,
        isSeen: Boolean,
    ) {
        transactionController.addTransaction(description, date, amount, isSeen)
    }

    private fun addTransactions(transactions: List<MomentumTransaction>) {
        transactionController.addTransactions(transactions = transactions)
    }

    fun deleteAllTransactions() {
        transactionController.deleteAllTransactions()
    }

    fun deleteTransactionById(transactionId: Int) {
        transactionController.deleteTransactionById(transactionId)
    }
}


